javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll

```diff
@@ -124,172 +124,243 @@ module AccessPath {
-    |
-      baseName = fromReference(write.getBase(), root)
+    /**
+     * Holds if the global `accessPath` is only assigned to from one file, not counting
+     * self-assignments.
+     */
+    cached
+    predicate isAssignedInUniqueFile(string accessPath) {
+      strictcount(File f | isAssignedInFile(accessPath, f)) = 1
+    }
+
+    /**
+     * Holds if `rhs` is directly assigned to the one-step global access path `path`.
+     *
+     * This predicate is mostly for internal use; most clients should use `getAnAssignmentTo`
+     * instead.
+     */
+    cached
+    predicate globalRhs(DataFlow::Node rhs, string path) {
+      exists(GlobalVariable var |
+        rhs = var.getAnAssignedExpr().flow() and
+        path = var.getName()
+      )
       or
-      baseName = fromRhs(write.getBase(), root)
-    )
-    or
-    exists(GlobalVariable var |
-      node = var.getAnAssignedExpr().flow() and
-      result = var.getName() and
-      root.isGlobal()
-    )
-    or
-    exists(FunctionDeclStmt fun |
-      node = DataFlow::valueNode(fun) and
-      result = fun.getId().(GlobalVarDecl).getName() and
-      root.isGlobal()
-    )
-    or
-    exists(ClassDeclStmt cls |
-      node = DataFlow::valueNode(cls) and
-      result = cls.getIdentifier().(GlobalVarDecl).getName() and
-      root.isGlobal()
-    )
-    or
-    exists(EnumDeclaration decl |
-      node = DataFlow::valueNode(decl) and
-      result = decl.getIdentifier().(GlobalVarDecl).getName() and
-      root.isGlobal()
-    )
-    or
-    exists(NamespaceDeclaration decl |
-      node = DataFlow::valueNode(decl) and
-      result = decl.getId().(GlobalVarDecl).getName() and
+      exists(FunctionDeclStmt fun |
+        rhs = DataFlow::valueNode(fun) and
+        path = fun.getId().(GlobalVarDecl).getName()
+      )
+      or
+      exists(ClassDeclStmt cls |
+        rhs = DataFlow::valueNode(cls) and
+        path = cls.getIdentifier().(GlobalVarDecl).getName()
+      )
+      or
+      exists(EnumDeclaration decl |
+        rhs = DataFlow::valueNode(decl) and
+        path = decl.getIdentifier().(GlobalVarDecl).getName()
+      )
+      or
+      exists(NamespaceDeclaration decl |
+        rhs = DataFlow::valueNode(decl) and
+        path = decl.getId().(GlobalVarDecl).getName()
+      )
+    }
+
+    /**
+     * Holds if `rhs` is assigned to the access path `ref` refers to, but with `step` appended.
+     *
+     * This predicate is mostly for internal use.
+     */
+    cached
+    predicate ref2rhs(DataFlow::Node ref, DataFlow::Node rhs, string step) {
+      any(DataFlow::PropWrite pw).writes(ref, step, rhs)
+    }
+
+    /**
+     * Holds if `succ` is assigned to the same access path as `pred`, but with `step` appended.
+     *
+     * This predicate is mostly for internal use.
+     */
+    cached
+    predicate rhs2rhs(DataFlow::Node pred, DataFlow::Node succ, string step) {
+      any(DataFlow::PropWrite pw).writes(pred, step, succ)
```

Comment:
This predicate seems to be identical to `ref2rhs`.

Reply:
Ah, yes; I suppose that isn't really very useful. I'll remove it.